tests/launcher.py

```diff
@@ -0,0 +1,709 @@
+                for keyObj in node['keys']:
+                    self.network.nodes[nodeName].keys.append(KeyStrings(keyObj['pubkey'], keyObj['privkey']))
+                for peer in node['peers']:
+                    self.network.nodes[nodeName].peers.append(peer)
+                for producer in node['producers']:
+                    self.network.nodes[nodeName].producers.append(producer)
+
+    def launch(self, instance: nodeDefinition):
+        dd = Path(instance.data_dir_name)
+        out = dd / 'stdout.txt'
+        err_sl = dd / 'stderr.txt'
+        err = dd / Path(f'stderr.{self.launch_time}.txt')
+        pidf = dd / Path(f'{Utils.EosServerName}.pid')
+
+        if instance.index in self.args.spcfc_inst_nums:
+            eosdcmd = [f"{getattr(self.args, f'spcfc_inst_{Utils.EosServerName}es')[self.args.spcfc_inst_nums.index(instance.index)]}"]
+        else:
+            eosdcmd = [Utils.EosServerPath]
+        if self.args.skip_signature:
+            eosdcmd.append('--skip-transaction-signatures')
+        if getattr(self.args, Utils.EosServerName):
+            eosdcmd.extend(shlex.split(getattr(self.args, Utils.EosServerName)))
+        if instance.index in self.args.specific_nums:
+            i = self.args.specific_nums.index(instance.index)
+            specifics = getattr(self.args, f'specific_{Utils.EosServerName}es')[i]
+            if specifics[0] == "'" and specifics[-1] == "'":
+                eosdcmd.extend(shlex.split(specifics[1:-1]))
+            else:
+                eosdcmd.extend(shlex.split(specifics))
+        eosdcmd.append('--config-dir')
+        eosdcmd.append(str(instance.config_dir_name))
+        eosdcmd.append('--data-dir')
+        eosdcmd.append(str(instance.data_dir_name))
+        eosdcmd.append('--genesis-json')
+        eosdcmd.append(f'{instance.config_dir_name}/genesis.json')
+        if self.args.timestamp:
+            eosdcmd.append('--genesis-timestamp')
+            eosdcmd.append(self.args.timestamp)
+
+        # Always enable a history query plugin on the bios node
+        if instance.name == 'bios':
+            if self.args.is_nodeos_v2:
+                eosdcmd.append('--plugin')
+                eosdcmd.append('eosio::history_api_plugin')
+                eosdcmd.append('--filter-on')
+                eosdcmd.append('"*"')
+            else:
+                eosdcmd.append('--plugin')
+                eosdcmd.append('eosio::trace_api_plugin')
+
+        if 'eosio::history_api_plugin' in eosdcmd and 'eosio::trace_api_plugin' in eosdcmd:
+            eosdcmd.remove('--trace-no-abis')
+            eosdcmd.remove('--trace-rpc-abi')
+            i = eosdcmd.index('eosio::trace_api_plugin')
+            eosdcmd.pop(i)
+            i -= 1
+            eosdcmd.pop(i)
+
+        if not instance.dont_start:
+            Utils.Print(f'spawning child: {" ".join(eosdcmd)}')
+
+            stdout = open(out, 'w')
+            stderr = open(err, 'w')
```

Comment:
What happens if `dd` hasn't been created yet?  Consider this change?

```suggestion
            os.makedir(dd, exist_ok=True)

            stdout = open(out, 'w')
            stderr = open(err, 'w')
```

Reply:
We considered it, rejected it, and now I'm reconsidering it.  Clayton suggested it and now you are and I think you're both right, just to deal with the slightly screwy position we find ourselves in in the name of backwards compatibility.